### PR TITLE
[CP] Add attention_mask to the buffer when the mask is causal 

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

* Ensured that the `attention_mask` is always checked for causality before being added to the buffer, and that this validation is performed only once for performance.
* Modified the logic so that the `attention_mask` is appended to buffers only after successful validation, preventing non-causal masks from being used.

### CODE REVIEW

**Review by SunMarc — APPROVED** [review]
Thanks !

**Comment by SunMarc** [comment]
cc @S1ro1 for visibility

**Comment by S1ro1** [comment]
It's useless to add this to buffers no? CP doesn't work with explicitly passed attention mask so in accelerate we attach a hook that pops it out

**Comment by kashif** [comment]
i was getting errors in the SFT Trainer since we use the  attention mask for metrics and its shape was not matching

**Comment by kashif** [comment]
@S1ro1  the entropy metric for example is erroring out:

https://github.com/huggingface/trl/blob/main/trl/trainer/sft_trainer.py#L1041-L1056

**Comment by S1ro1** [comment]
Okay, ig makes sense. But in this case it's not entirely correct to shard mask across dim(1), as mask should be sharded across dim(1) and dim(2) which is not expressable in torch DTensor placements, so just to look after that.

**Comment by kashif** [comment]
@S1ro1  good point! How about we only split the 2d masks?
